Canvas context type for WebGPU

toji opened the pull request:
<!--
Thank you for contributing to the HTML Standard! Please describe the change you are making and complete the checklist below if your change is not editorial.
-->

- [x] At least two implementers are interested (and none opposed):
   * Chrome
   * Edge
   * Firefox
   * Safari
   Discussion: https://github.com/gpuweb/gpuweb/issues/131
- [x] Tests are written and can be reviewed and commented upon at:
   * WebGPU has it's [own set of conformance tests](https://github.com/gpuweb/cts), which is evolving along with the WebGPU spec itself.
   * Tests for this specific area of the spec are [located here](https://github.com/gpuweb/cts/tree/main/src/webgpu/web_platform/canvas).
- [x] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chrome: https://bugs.chromium.org/p/chromium/issues/detail?id=852089
   * Firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=1602129
   * Safari: https://bugs.webkit.org/show_bug.cgi?id=167952

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)

---------------

Posting this as a PR so that the proposed changes can be reviewed, but before landing the WebGPU working group would like to discuss the proposed context type string, `"gpupresent"`, with this group. The name differs somewhat from other context types and we want to ensure that it fits the desired patterns for these strings (if there is any) and that the name is appropriately suggestive of it's function. See https://github.com/gpuweb/gpuweb/issues/131 for prior discussion within the WebGPU working group on the topic.

As a brief overview: As opposed to the WebGL contexts, the `"gpupresent"` context type (which would return a [`GPUPresentationContext`](https://gpuweb.github.io/gpuweb/#presentation-context) object) is not the primary interface for interacting with the GPU hardware in WebGPU API. That's the responsibility of the [`GPUDevice`](https://gpuweb.github.io/gpuweb/#gpu-device) which is created from a [`GPUAdapter`](https://gpuweb.github.io/gpuweb/#gpu-adapter) enumerated by the `navigator` interface. This is to accommodate non-visual uses of the API (making use of GPU hardware for compute capabilites, for example) and also to allow a single device to draw to multiple canvas elements.

As a result, the `"gpupresent"` canvas context primarily provides an surface for `GPUDevice` to draw to. Because of this the WebGPU working group generally wants to avoid names like `"webgpu"`, `"gpu"`, or anything that could be mistaken as returning the primary WebGPU interface used for rendering. `"bitmaprenderer"` has a somewhat similar function to the `"gpupresent"` context, in that it only displays images provided to it rather than producing the images itself. It's possible that a context named `"gpurenderer"` could imply that the context was used to produce (render) the image, however, so the term "present" is preferred here as being less ambiguous.

A final point of discussion within the WebGPU group is whether or not the context names should be considered analogous enums, and thus prefer using hyphenation? (ie: `"gpu-present"`) It seems that `"bitmaprenderer"` has already established a precedent for not hyphenating, which is why `"gpupresent"` was chosen, but we are happy to help establish a new precedent going forward if this group would prefer.


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***

### :boom: Error: Wattsi server error :boom: ###

[PR Preview](https://github.com/tobie/pr-preview#pr-preview) failed to build. _(Last tried on Jul 14, 2021, 7:04 PM UTC)_.

<details>
<summary>More</summary>


PR Preview relies on a number of web services to run. There seems to be an issue with the following one:

:rotating_light: [Wattsi Server](https://github.com/domenic/wattsi-server) - Wattsi Server is the web service used to build the WHATWG HTML spec.

:link: [Related URL](https://build.whatwg.org/wattsi)

```
<html>
<head><title>524 Origin Time-out</title></head>
<body bgcolor="white">
<center><h1>524 Origin Time-out</h1></center>
<hr><center>cloudflare-nginx</center>
</body>
</html>

```

_If you don't have enough information above to solve the error by yourself (or to understand to which web service the error is related to, if any), please [file an issue](https://github.com/tobie/pr-preview/issues/new?title=Error%20not%20surfaced%20properly&body=See%20whatwg/html%236804.)._
</details>

kainino0x commented:
@Juanmihd FYI

annevk commented:
@toji did you intentionally mark this as WIP? You'll need to make your membership of the googlers GitHub organization public for the Participation check.

toji commented:
I did mark it as WIP intentionally to underscore the fact that we want to discuss the `"gpupresent"` string prior to landing.

My organization memberships are public, but I don't seem to be in the Google org for some reason? I thought I was. I'll look into that.

domenic commented:
I see you've joined the org, but your membership is still not public (you can check by viewing in private browsing mode) :).

I'll join the discussion in https://github.com/gpuweb/gpuweb/issues/131 for the naming question.

domenic commented on source:
```diff
@@ -60305,13 +60324,16 @@ callback <dfn callback>BlobCallback</dfn> = undefined (<span>Blob</span>? blob);
-    data-x="canvas-context-webgl2">webgl2</code>" contexts. <ref spec=WEBGL></p>
+    data-x="canvas-context-webgl2">webgl2</code>" contexts. <ref spec=WEBGL> The WebGPU
+    specification defines the "<code data-x="canvas-context-gpupresent">gpupresent</code>" context.
+    <ref spec=WEBGPU></p>
```
Nit: we put the `<ref>`s at the end of the paragraph together, instead of interspersed among the sentences.

domenic commented on source:
```diff
@@ -60405,6 +60432,29 @@ callback <dfn callback>BlobCallback</dfn> = undefined (<span>Blob</span>? blob);
+       <th>"<dfn><code data-x="canvas-context-gpupresent">gpupresent</code></dfn>", if the user
+       agent supports the WebGPU feature in its current configuration
+       <td>
+        Follow the instructions given in the WebGPU specification's <i>Canvas Rendering</i> section
```
Can we make this an actual `<a>` link?

domenic commented on source:
```diff
@@ -66119,15 +66171,18 @@ interface <dfn interface>OffscreenCanvas</dfn> : <span>EventTarget</span> {
-    data-x="offscreen-context-type-webgl2">webgl2</code>" contexts. <ref spec=WEBGL></p>
+    data-x="offscreen-context-type-webgl2">webgl2</code>" contexts. <ref spec=WEBGL> The WebGPU
+    specification defines the "<code data-x="offscreen-context-type-gpupresent">gpupresent</code>"
+    context. <ref spec=WEBGPU></p>
```
refs at end of paragraph as above

domenic commented on source:
```diff
@@ -66295,6 +66356,28 @@ interface <dfn interface>OffscreenCanvas</dfn> : <span>EventTarget</span> {
+      <tr>
+       <th>"<dfn><code data-x="offscreen-context-type-gpupresent">gpupresent</code></dfn>"
+       <td>
+        Follow the instructions given in the WebGPU specification's <i>Canvas Rendering</i> section
```
Hyperlink as above, if possible

domenic reviewed (commented):
Spec text LGTM with nits

toji commented:
Thanks for the feedback! Updated the PR with the fixes.

toji commented:
Thank you all for the feedback we received on https://github.com/gpuweb/gpuweb/issues/131! As a result we've updated the context type and interface names used to take into account some of the concerns that were raised. Due to potential confusion around the term "present" or "presentation" we've dropped that verb and now define `'webgpu'` as the context type (rather than `'gpupresent'`) and the interface it return will be a `GPUCanvasContext` (rather than a `GPUPresentationContext`), both of which should be more closely inline with existing context variants.

The functionality exposed by this interface has not changed, and it still functions as a way of returning drawing surfaces to the developer rather than the API which performs the drawing.

This pull request has been updated to reflect those changes, and as a result has been taken out of Draft mode. Please take a look and let us know if any further updates are required!

domenic reviewed (approved):
LGTM; thanks for the great PR!
